### PR TITLE
Java: SSRF query

### PR DESCRIPTION
**Pull request by luchua-bc**
Server Side Request Forgery (SSRF) is an attack vector that abuses an application to interact with the internal/external network or the machine itself, frequently through the attack vector of URL mishandling.

In an SSRF attack, the attacker can abuse functionality on the server to read or update internal resources, e.g. cloud server meta-data, internal REST interface, and protected files.

This query detects the following two scenarios:</p>
+ Java networking uri.openConnection() and its derived uri.openStream(), which is a shorthand for openConnection().getInputStream(), from remote source
+ Apache HTTPRequest constructed from remote source</li>

There are two existing queries in the CodeQL repository that handle two important/specific cases of this vulnerability category:
+ [CWE-319](https://github.com/github/codeql/blob/d9bc07cb912e8acfaec273f43e6dcfa3230c463d/java/ql/src/Security/CWE/CWE-319/UseSSL.ql) for cleartext HTTP connection
+ [CWE-036](https://github.com/github/codeql/blob/01157e43e35b4e8efb2d2c9e2fb875150bf7f637/java/ql/src/experimental/Security/CWE/CWE-036/OpenStream.ql) for the special case of openConnection().getInputStream() leading to local file disclosure

Thus this query has a broader scope and also covers the frequently used Apache HTTP library.

Please consider to merge the PR. Thanks.

### CODE REVIEW

File: java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+<qhelp>
+
+  <overview>
+    <p>Server Side Request Forgery (SSRF) is an attack vector that abuses an application to interact with the internal/external network or the machine itself, frequently through the attack vector of URL mishandling.</p>
+    <p>In an SSRF attack, the attacker can abuse functionality on the server to read or update internal resources, e.g. cloud server meta-data, internal REST interface, and files.</p>
+    <p>By providing unexpected URLs, attackers can trick the application to send malicious requests, possibly bypassing access controls such as firewalls that prevent the attackers from accessing those URLs directly. The server can be used as a proxy to conduct port scanning of hosts in internal networks, use other URLs such as that can access documents on the system (using file:// or jar://), or use other content protocols such as gopher:// or tftp://, or invoke internal REST APIs.</p>
+    <p>This query detects the following two scenarios:</p>
+    <ol>
+      <li>Java networking uri.openConnection() and its derived uri.openStream(), which is a shorthand for openConnection().getInputStream(), from remote source.</li>
```
Comment (Marcono1234) [inline java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp:10]:
```suggestion
      <li>Java networking URL.openConnection() and its derived URL.openStream(), which is a shorthand for openConnection().getInputStream(), from remote source.</li>
```
To make it more clear to which class the methods belong

Might also be good to rename the local variables in the code examples from `uri` -> `url`

File: java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    <p>The following example shows both 'BAD' and 'GOOD' configurations. In the 'BAD' configuration, remote source is not validated before sending requests. In the 'GOOD' configuration, remote source is validated before sending requests.</p>
+    <sample src="SSRF.java" />
+  </example>
+
+  <references>
+    <li>
+      <a href="https://cwe.mitre.org/data/definitions/918.html">CWE-918: Server-Side Request Forgery (SSRF)</a>
+      <a href="https://owasp.org/www-community/attacks/Server_Side_Request_Forgery">OWASP - Server Side Request Forgery</a>
+      <a href="https://cheatsheetseries.owasp.org/cheatsheets/Server_Side_Request_Forgery_Prevention_Cheat_Sheet.html">Server Side Request Forgery Prevention</a>
+      <a href="https://www.ciphertechs.com/blog/hawtio-advisory/">CVE-2019-9827:  Unauthenticated Server-Side Request Forgery (SSRF) with Hawtio up to and including version 2.5.0</a>
```
Comment (Marcono1234) [inline java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp:29]:
```suggestion
      <a href="https://www.ciphertechs.com/blog/hawtio-advisory/">CVE-2019-9827: Unauthenticated Server-Side Request Forgery (SSRF) with Hawtio up to and including version 2.5.0</a>
```
One space too much

File: java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp (hunk 1 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp]: Thanks @Marcono1234 . This is a very good point. I've added a sanitizer and relevant test cases to reduce false positives. And I've made text changes as well.

File: java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp (hunk 2 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-918/SSRF.qhelp]: Thanks @Marcono1234 . Change was made and committed.

**Comment by smowton** [comment]
Hello, apologies for our slow response on this: we've been backlogged with requests and are just now clearing the queue.

Because this PR has the same topic as #3454 but was both submitted later and has a smaller feature-set (e.g., the other PR supports a broader range of HTTP client libraries), I'm sorry to say we can't accept this submission.

If you are able to improve upon that PR, please do submit changes based on top of that branch, and we will be happy to assess their contribution relative to the baseline.

**Pull request closed without merging**